traitlets/config/configurable.py

```diff
@@ -203,8 +210,13 @@ def _config_changed(self, change):
         section_names = self.section_names()
         self._load_config(change.new, traits=traits, section_names=section_names)
 
-    def update_config(self, config):
-        """Update config and load the new values"""
+    def update_config(self, config, skip_env=False):
```

Comment:
Another proof that compatibility does not suffer under this PR is that ALL pre-existing TCs passed this PR without further modifications - any TC-changes were only to strengthen this new feature.

So the remaining concern which is the "correct" order.
I vote for *configs-->env-->cmdline*, so bash-commands like this run as expected:

    JUPYTER_HOME=/opt/foo jupyter notebook

and not having to worry about clashes with config-file.
And **`update_config()` should enforce this rule by default**, UNLESS some exceptional reason exists, i.e. configs sourced from cmd-line args or user-locals.

Choosing the opposite defaults, and hoping that downstream code will *somehow* achieve the above order, is too "optimistic".  

Note-1: I did some corrections in my previous comment - please read it from GH.
Note-2: I can address your points one-by-one, @minrk, if you think it would facilitate discussion.